numpy/_core/src/multiarray/ctors.c

```diff
@@ -1829,11 +1829,14 @@ PyArray_CheckFromAny_int(PyObject *op, PyArray_Descr *in_descr,
 {
     PyObject *obj;
     if (requires & NPY_ARRAY_NOTSWAPPED) {
-        if (!in_descr && PyArray_Check(op)) {
-            in_descr = PyArray_DESCR((PyArrayObject *)op);
-            Py_INCREF(in_descr);
+        if (!in_descr && PyArray_Check(op) &&
+                PyArray_ISBYTESWAPPED((PyArrayObject* )op)) {
+            in_descr = PyArray_DescrNew(PyArray_DESCR((PyArrayObject *)op));
+            if (in_descr == NULL) {
+                return NULL;
+            }
         }
-        if (in_descr) {
+        else if (in_descr && !PyArray_ISNBO(in_descr->byteorder)) {
             PyArray_DESCR_REPLACE_CANONICAL(in_descr);
```

Comment:
Should we go all the way back and not use `PyArray_DESCR_REPLACE_CANONICAL` for backport?
(Basically, just adding the extra `!PyArray_ISNBO(in_descr->byteorde`).

But if not, this looks fine.  It'll solve the crash-regression for sure, thanks!

Reply:
I think that makes more sense to do to the 2.2.x branch than to `main`. I'd rather use `ensure_canonical` in more places. I'll just go ahead and make another PR to the stable branch with the minimal change, just in case there are other structured dtype breakages that didn't get so quickly reported.

Reply:
See https://github.com/numpy/numpy/pull/28201

Reply:
OK, but that means you need to make a different/dedicated backport PR against `2.2.x`?  (If I think about it again, for `main` we could potentially keep the code exactly as is, the bug may be only in `searchsorted` (the bug being that the `dtype` passed in may be replaced, so one cannot keep using it).

Reply:
Ah, thanks for the backport.  I'll have another look at searchsorted, maybe we can just fix it up there (and maybe add a brief release note just in case).

Reply:
Great! Feel free to close this if you come up with an alternate fix.